Abstraction of OpenGL Data types

Quaker762 opened the pull request:
Very basic abstraction of more complex OpenGL concepts (Vertex Array Objects and Vertex Buffer Objects). Makes the whole thing _much_ more readable and easier to handle. Should also make everything a little bit
more typesafe and memory safe (no segfaults).

There's also a nice demo in main as to how it works. As is clearly visible, there is a LOT less OpenGL bullshit to type out, as the object does it all for you! :)

Quaker762 commented:
Oh, and I've discovered possibly the best thing ever. It's called CodeXL (it's an OpenGL/OpenCL/CPU Debugger). I'll update the readme, as _everyone_ should be using this to debug 3D data (especially when we get to anything cutscene releated).

![debug](https://user-images.githubusercontent.com/4768900/27145089-ae8b3fb6-5177-11e7-9156-a34a95da6ecc.png)

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+     * An OpenGL Vertex Buffer Object (VBO). Provides a nice, easy, typesafe way
+     * to create and manage OpenGL buffers.
+     */
+    struct sh3_glbuffer final
```
The name seems redundant with the namespace. I think this should just be `buffer`.

Quaker762 replied:
Yeah good idea, also considering this technically isn't a part of the actual game. I'll change it.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+    {
+    public:
+
+        enum Target
```
Any reason why this is not an `enum class`?

Quaker762 replied:
I'm not quite sure of the difference, could you possibly explain? All Google is telling me is that it's more strongly typed.

z33ky replied:
Yes, that is one difference. This means there's no implicit cast to or from numeric types (`int x = SH3_ARRAY_BUFFER; Target y = 5;`). This is still possible, it just needs explicit casts.
And scoping is enforced, so the `Target::` prefix is required.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+         */
+        sh3_glbuffer(){Create();buffType = Target::SH3_ARRAY_BUFFER;};
+        sh3_glbuffer(Target type){Create();buffType = type;};
+        sh3_glbuffer(Target type, const std::string& buffname){Create();buffType = type, name = buffname;};
```
You could provide default values for those parameters to get rid of the overloads.
Also, is it important that `Create()` is called before initializing the members? Otherwise you can use initializer lists.

Quaker762 replied:
I'll get on that ASAP. There's no reason, in fact I should have done this in the reverse order.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+        sh3_glbuffer(Target type, const std::string& buffname){Create();buffType = type, name = buffname;};
+
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted by the C++ Runtime.
```
"by the C++ Runtime"??
Anyway, seems like an redundant statement, since "Destructor" should imply as much.

Quaker762 replied:
> "by the C++ Runtime"??

Doesn't the runtime library handle all this? Regardless I'll remove it.

z33ky replied:
That'd be true if we'd be using a garbage collected language.
Perhaps you could say that using a smart pointer would also qualify, since the standard library includes the code to invoke the destructor.
Usually however, destructor calls are inserted at the end of scopes by the compiler, requiring no C++ runtime to even be present.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted by the C++ Runtime.
+         */
+        ~sh3_glbuffer(){/**Release();*/};
```
Why is this commented out?

Quaker762 replied:
I'm not quite sure how OpenGL works on this instance. Release frees the buffer's ID, and I'm not entirely sure if the data is released with it. If we create the buffers locally in say `sh3_model::Load( )`, and then release is called when the function returns, the data should be deleted from the GPU (and all the buffers in buffer becoming invalid)? I'll need to a bit more investigation, hence the commenting. I'll remove the line completely for now.

z33ky replied:
Seems to work, if we want to trust [this](https://stackoverflow.com/a/19645105) StackOverflow answer.

Quaker762 replied:
Interesting. The information for this is so fragmented, thanks for finding this haha.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+         *  @param num Size (in bytes) of the data we are flushing.
+         *  @param data void pointer to our data (usually an std::array or std::vector).
+         */
+        void BufferData(std::size_t num, void* data);
```
I didn't actually check if this is indeed common, but I think most APIs pass the data first and then the length. I'd like to stay consistent if there's a rule.
Also, I think it'd make sense to provide an API like `sh3_arc_file::ReadObject` which uses templates to infer the size.

Quaker762 replied:
Okay I'll get on it. Need to study up a bit on templates.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,112 @@
+         *  @return Number of elements in this GL Buffer.
+         */
+        std::uint32_t GetNumberElements() const;
+        std::string     name; /**< The name of this buffer */
```
Any reason to have this not be `const`?

z33ky commented on include/SH3/system/glprogram.hpp:
```diff
@@ -48,9 +48,11 @@ namespace sh3_graphics
         private:
         };
 
+        sh3_glprogram(const std::string& name) : programName(name){Load(name);}
```
I'm not sure if I'm a fan of this overload. The version below still allows to ignore errors, but makes it visible to the caller that there could be one that can be checked for.

Quaker762 replied:
Fixed, I had a bit of brainfart in my rush to implement this.

z33ky commented on include/SH3/system/glprogram.hpp:
```diff
@@ -48,9 +48,11 @@ namespace sh3_graphics
         sh3_glprogram(const std::string& name, load_error& err, const std::vector<std::string>& attribs = {}) : programName(name){Load(name, err, attribs);};
         ~sh3_glprogram(){Unbind(); glDeleteProgram(programID);}
 
+        void Load(const std::string& name);
```
See above.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,118 @@
+         *  sent to it from a VBO. These values were picked for no specific reason other than
+         *  the usual order of a vertex (vert data, normal data, uv data).
+         */
+        enum Attribute
```
Any reason why this is not an `enum class`?

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,118 @@
+        /**
+         *  Specifies data type of data passed to paramater 3 of glVertexAttribPointer.
+         */
+        enum DataType
```
Any reason why this is not an `enum class`?

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,118 @@
+         */
+        enum Attribute
+        {
+            SH3_VERTEX_SLOT = 0,
```
I think automatic numbering should be fine here.

z33ky commented on source/SH3/system/glbuffer.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void sh3_glbuffer::Create()
+{
+    GLenum err;
```
?

Quaker762 replied:
Leftover code I should have removed. `glGetErorr( )` is actually a VERY bad function. It returns an error from any last glXYZ call without being cleared. We could have an issue in another part of the code that would trigger the error handler to be called, even if our buffer creation was successful. I'll remove these.

z33ky replied:
We can define a `SH3ClearGLError` function, which pops all errors of the error queue and warns about them. But this should be in a separate commit. Even in a separate pull request if you want.

Quaker762 replied:
I'll get on that ASAP.

z33ky commented on source/SH3/system/glbuffer.cpp:
```diff
@@ -0,0 +1,52 @@
+#include "SH3/system/glbuffer.hpp"
+#include "SH3/system/log.hpp"
+
+#define GL_VBO_UNBIND 0
```
Use `constexpr` instead.

z33ky commented on source/SH3/system/glbuffer.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void sh3_glbuffer::Bind()
+{
+    GLenum err;
```
?

z33ky commented on source/SH3/system/glprogram.cpp:
```diff
@@ -48,6 +48,44 @@ std::string sh3_glprogram::load_error::message() const
 }
 
+
+void sh3_glprogram::Load(const std::string& name)
```
This should just call the other overload instead of duplicating the implementation.

Quaker762 replied:
I think OpenGL will crash if I specify no attributes, I'll have to double check. I need to research a bit more to better abstract some of this stuff.

z33ky replied:
Is this method not equivalent to the other Load-overload when called with an empty `attribs` vector?

Quaker762 replied:
It is. It's been a while since I've looked over the `glprogram` code, so I kinda forgot haha.

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+#include "SH3/system/glvertarray.hpp"
+#include "SH3/system/log.hpp"
+
+#define GL_VAO_UNBIND 0
```
Use `constexpr`.

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+#include "SH3/system/log.hpp"
+
+#define GL_VAO_UNBIND 0
+#define BUFFER_OFFSET(i) ((char *)NULL + (i))
```
Use `constexpr`.

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void sh3_glvertarray::Create()
+{
+    GLenum err;
```
?

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+    GLenum err;
+
+    glGenVertexArrays(1, &id);
+
```
Remove this line.

Quaker762 replied:
Remove `glGenVertexArrays(1, &id);`????

z33ky replied:
No, the empty line after that :)

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void sh3_glvertarray::Bind()
+{
+    GLenum err;
```
?

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void sh3_glvertarray::AddVBO(sh3_glbuffer& buff)
+{
+    if(buffers.size() + 1 == 4) // If we have 4 VBOs, we know for sure we have an indexed VAO (Slot 1 VERTS, Slot 2 NORMALS, Slot 3 UV, Slot 4 INDEX)
```
`if(buffers.size() == SH3_INDEX_SLOT)` makes the intent clear I think.

Quaker762 replied:
Wouldn't that imply that the next size would be 5, and the index buffer has already been added, hence crashing OpenGL?

EDIT: Nevermind, I forgot what section of code this was in.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,118 @@
+        GLuint                      id; /**< This VAO's ID. */
+        std::vector<sh3_glbuffer>   buffers; /**< All VBOs managed by this VAO. MUST BE IN ORDER SPECIFIED BY @ref Attribute. */
+
+        bool indexed = false;
```
Replace with `bool is_indexed() const { return buffers.size() == *; }`?

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,118 @@
+
+    private:
+        GLuint                      id; /**< This VAO's ID. */
+        std::vector<sh3_glbuffer>   buffers; /**< All VBOs managed by this VAO. MUST BE IN ORDER SPECIFIED BY @ref Attribute. */
```
`sh3_glbuffer` seems small enough that we could just store an `std::array` of sh3_glbuffers with the maximum possible number of buffers, which is also a reasonably small 4 (this constant could also be included in the `Attribute` `enum` as `MAX` or something.
We then need an additional member to keep track of which buffers have been initialized. I guess an `std::underlying_type_t<Attribute>` which gets incremented when calling AddVBO does the job.

z33ky commented on source/main.cpp:
```diff
@@ -32,17 +47,45 @@ int main(int argc, char** argv)
+
+    color.BufferData(sizeof(glCol), glCol);
+    vao.AddVBO(color);
+    vao.SetDataLocation(sh3_gl::sh3_glvertarray::Attribute::SH3_NORMAL_SLOT, sh3_gl::sh3_glvertarray::DataType::SH3_FLOAT, 3, 0, 0);
```
If we pass the color as normal data, the API seems off.

Quaker762 replied:
This was just to test if the data was being passed correctly to the shader.

z33ky reviewed (changes requested):
When changing the `enum`s to `enum class`es, then the `SH3_` prefix becomes unnecessary.

I'm not sold on the `AddVBO(); SetDataLocation();` API.
I think AddVBO should either return a handle, which can be passed to `SetDataLocation` instead of specifying the `Attribute`, or (preferably) a proxy-object, which you can call `SetDataLocation` on without having to specify any kind of `Attribute`.

Instead of using `vector.at(index)` I think we can just `assert(vector.size() >= index); vec[index]` to skip the bounds check in release-mode.

I'm not a huge fan of the way `main.cpp` is used to test whatever feature was most recently added.
It would make more sense to have each of this in a separate `main()` method in a separate file (and thus program).
However, this is not an issue with this pull request. We should open a new issue for this.

Quaker762 commented:
> I think AddVBO should either return a handle, which can be passed to SetDataLocation

What handle would I return from AddVBO?? The whole point of it I suppose is to have a list of all current buffers loaded for a particular VAO. 

> instead of specifying the Attribute, or (preferably) a proxy-object, which you can call SetDataLocation on without having to specify any kind of Attribute.

How would I go about doing this? The only issue I can see is that `glVertexAttribPointer` requires an attribute in parameter one. I suppose I could add that to the buffer itself, instead of having an enum list. How would I pass a proxy object in??? IMHO it's easier to just have a list of the buffers and bind them, due to the way OpenGL works. Mainly I have it like this because it adheres to the way OpenGL requires us to do things without the annoyance of cluttering up the code with glXYZ calls everywhere and raw data manipulation. 

>I'm not a huge fan of the way main.cpp is used to test whatever feature was most recently added.
It would make more sense to have each of this in a separate main() method in a separate file (and thus program).

I probably should have just taken a screenshot, but I kind of wanted to demonstrate how the API for this works, as it was a little bit fiddly, and there are certain (calling) rules that need to be adhered to so that OpenGL doesn't shit the bed. 

Would it be better to have a whole lot of C functions then, rather than implementing this in classes??

z33ky commented:
> What handle would I return from AddVBO?? The whole point of it I suppose is to have a list of all current buffers loaded for a particular VAO.

And you would keep that. AddVBO would just return a handle, like the index of the most recently added object, to identify the VBO that was added. This is used by `SetDataLocation` to identify the object.
In fact, the index already exists, its value just assumed to be whatever `enum Attribute` constant.

> How would I go about doing this? The only issue I can see is that `glVertexAttribPointer` requires an attribute in parameter one. I suppose I could add that to the buffer itself, instead of having an enum list. How would I pass a proxy object in??? IMHO it's easier to just have a list of the buffers and bind them, due to the way OpenGL works. Mainly I have it like this because it adheres to the way OpenGL requires us to do things without the annoyance of cluttering up the code with glXYZ calls everywhere and raw data manipulation.

Something like this:
```
struct ProxyObject final
{
    Attribute self;
    sh3_glvertarray &array;
    void SetDataLocation(...) { array.SetDataLocation(self, ...); }
};
```
I think you misunderstood what I had in mind with a proxy object.

> I probably should have just taken a screenshot, but I kind of wanted to demonstrate how the API for this works, as it was a little bit fiddly, and there are certain (calling) rules that need to be adhered to so that OpenGL doesn't shit the bed.

It makes sense, both to illustrate the API like you stated, and also to have a test ready in case someone wants to change something and test that it still works.
I don't think removing/commenting out whatever was last in `main()` is a good approach though, since the next thing to test is going to come along and `main()` will be cluttered up some more (or if the code is removed, the example/test case is gone).

> Would it be better to have a whole lot of C functions then, rather than implementing this in classes??

I didn't state that the OO encapsulation was bad. In contrast, I think it's a good approach. I just don't like the implicit assumptions about the `sh3_vertarray::buffers`-index (i.e. the values of `enum Attribute` to use).

Quaker762 commented:
> And you would keep that. AddVBO would just return a handle, like the index of the most recently added object, to identify the VBO that was added. This is used by SetDataLocation to identify the object.
In fact, the index already exists, its value just assumed to be whatever enum Attribute constant.

Ohh I see, that's a much better way of handling it, I didn't think of that.

> I think you misunderstood what I had in mind with a proxy object.

So basically something that encapsulates the VAO without having to pass an attribute? I'm not quite sure that will work with what I have planned for the 3D model section.

> I don't think removing/commenting out whatever was last in main() is a good approach though, since the next thing to test is going to come along and main() will be cluttered up some more (or if the code is removed, the example/test case is gone).

I see, I have the same sentiment. Are there any testing frameworks we can use (like jUnit), or is better to have multiple mains in a `test` folder and then have test targets?? Might come in handy for a 3D camera test.

> I just don't like the implicit assumptions about the sh3_vertarray::buffers-index (i.e. the values of enum Attribute to use).

My reasoning is that (at least from my deductions), that it's up the programmer to decide which attributes are bound where. For example, we could have it in the order: UV, vertex, normals, but that's a pretty bad idea. It's also not _really_ an implicit assumption per se, as SILENT HILL's vertex format is in this **exact** same order, so it's technically consistent. But I do agree, the handle idea does make it a lot cleaner and makes a lot more sense, though an attribute ID is technically different to an attribute index. I've gone and changed it to the handle idea, but I still think having a weakly typed regular `enum` to pass to OpenGL is a good way to handle it.

Quaker762 commented:
I've just stumbled upon another GL abstraction library here: https://github.com/Overv/OOGL/blob/150aebfa307a76119f62085cb0d73991793c0dbe/include/GL/GL/VertexArray.hpp

It seems that this author came to almost the same conclusion I did for his VAOs and VBOs (I'm taking a look at other C++ abstractions of OpenGL data objects)

z33ky commented:
> So basically something that encapsulates the VAO without having to pass an attribute? I'm not quite sure that will work with what I have planned for the 3D model section.

`Attribute a = vao.AddVBO(o); vao.SetDataLocation(a, ...);` is equivalent to `Proxy p = vao.AddVBO(o); p.SetDataLocation(...)`.

> I see, I have the same sentiment. Are there any testing frameworks we can use (like jUnit), or is better to have multiple mains in a `test` folder and then have test targets??

There is no one big C++ testing framework. A colleague of mine recently reviewed testing frameworks of various open source C libraries, I'll ask him if he knows some good C++ ones.

To start with, I think just a test-directory with one target (executable) per test and no automatic testing is fine.

> My reasoning is that (at least from my deductions), that it's up the programmer to decide which attributes are bound where. For example, we could have it in the order: UV, vertex, normals, but that's a pretty bad idea. It's also not _really_ an implicit assumption per se, as SILENT HILL's vertex format is in this **exact** same order, so it's technically consistent.

The assumption being backed by the data layout does not make it explicit.
In the current version, the attribute that gets modified by a call to `SetDataLocation` is not visible in the call itself, it can only be implicitly from the context where (and when) it is called.

> But I do agree, the handle idea does make it a lot cleaner and makes a lot more sense, though an attribute ID is technically different to an attribute index.

Right. If I understand this right, an attribute ID is a higher level concept represented by the `enum Attribute`, right? While the attribute index is the index into `sh3_glvertarray::buffers`?
And it "just happens" that there's a 1-to-1 mapping from ID to index, yeah?

Anyways, if this change is adopted then the `enum Attribute` is not even needed, at least not in this class. It can instead be moved to the `class texture`.
We might also want to use this VAO abstraction for procedurally generated objects, e.g. quads, boxes or spheres used for debugging, where we might only have vertices (SH3_VERTEX_SLOT) and indices (SH3_INDEX_SLOT), but no texture. So there we would use a different enumeration.

Quaker762 commented:
> Attribute a = vao.AddVBO(o); vao.SetDataLocation(a, ...); is equivalent to Proxy p = vao.AddVBO(o); p.SetDataLocation(...).

Ohhh okay I get it now.

> To start with, I think just a test-directory with one target (executable) per test and no automatic testing is fine.

Yeah this sounds much better. Would it perhaps be feasible to compile the engine as libsh3.a or something like that (basically an engine/game detachment). I'll open an issue if you think this might work.

> The assumption being backed by the data layout does not make it explicit.
In the current version, the attribute that gets modified by a call to SetDataLocation is not visible in the call itself, it can only be implicitly from the context where (and when) it is called.

I think I see what you mean now, so the caller knows what type was passed in, but the actual _function_ itself doesn't know what type was explicitly passed (because of the weak typing)?

> Right. If I understand this right, an attribute ID is a higher level concept represented by the enum Attribute, right? While the attribute index is the index into sh3_glvertarray::buffers?
And it "just happens" that there's a 1-to-1 mapping from ID to index, yeah?

Yeah, it's sort of hard to explain. At the end of the day, it's all got to do with _how_ the streams are passed into the vertex shader so we can perform operations on them (using the `layout(location = 0) in` style syntax). It makes more sense to hardcode the positions, because, as you've said, we might bind a quad with only vertex index and uv data, which makes our shaders consistent as well.

> Anyways, if this change is adopted then the enum Attribute is not even needed, at least not in this class. It can instead be moved to the class texture.

How so? The attributes are part of the VAO (even part of the glprogram is you can spin it like that I suppose haha). 

> We might also want to use this VAO abstraction for procedurally generated objects, e.g. quads, boxes or spheres used for debugging, where we might only have vertices (SH3_VERTEX_SLOT) and indices (SH3_INDEX_SLOT), but no texture. So there we would use a different enumeration.

Yes!! This is point of abstracting it the way I did. Not all geometry will be laid out the same way, and having (if you will) "pre-baked" locations makes it _much_ simpler imho when it comes to writing a shader, as we know what to expect in which location. Although, I do see what you mean by having the attributes per class (like having them in texture so we can map vertex to 0, uv to 1 and index to 2 etc, which will work given the current implementation).

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,109 @@
+        /**
+         *  Default constructor.
+         */
+        sh3_buffer(Target type, const std::string& buffname = ""){buffType = type, name = buffname;Create();};
```
Use initializer lists (`(... buffname = ""):bufftype(type), name(buffname)`).

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,109 @@
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted by the C++ Runtime.
+         */
+        ~sh3_buffer(){/**Release();*/};
```
This should be fine, right? Otherwise put a `//FIXME` there explaining why this is not done currently.

Quaker762 replied:
I fixed this, stashed it and then forgot to unstash.. 

Sorry haha.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,109 @@
+         *  @param num Size (in bytes) of the data we are flushing.
+         *  @param data void pointer to our data (usually an std::array or std::vector).
+         */
+        void BufferData(void* data, std::size_t num);
```
`sh3_arc_file::ReadObject`-like interface? You should be able to just copy it over and adapt the parameters, even if you're not up to speed on templates.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,109 @@
+    private:
+        GLuint          id;         /**< The ID of our buffer given to us by OpenGL */
+        Target          buffType;   /**< What type of buffer this is */
+        std::uint32_t   size;      /**< The size of this buffer in bytes */
```
Misaligned comment :S

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,126 @@
+         */
+        enum class Attribute
+        {
+            VERTEX_SLOT = 0,
```
`enum`s start with 0 if you don't specify it.

Quaker762 replied:
Ahhh good I wasn't entirely sure if it didn't without specifying at least one.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,126 @@
+         */
+        bool is_indexed() const
+        {
+            return buffers.size() == 4;
```
Should be `== static_cast<std::underlying_type_t<Attribute>(INDEX_SLOT)`.

z33ky reviewed (changes requested):
You skipped many of the requested improvements :(
Why is that?

Quaker762 commented:
> You skipped many of the requested improvements :(
Why is that?

Aside from what I emailed you about, I _maaaay_ have forgotten to do a git stash pop at 12am sorry HAHA.

Quaker762 commented:
@z33ky 

I've changed the way this works to your 'handle' idea. It actually works much better than how I originally had it, and also makes sure correct order is maintained across the VAO and shader. It should work really well, but until we have some real OpenGL up and running, it's hard to tell haha.

Quaker762 commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,120 @@
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted by the C++ Runtime.
+         */
+        ~sh3_buffer(){/**Release(); FIXME: This causes a segmentation violation (why is this called in the game loop!?!?!)*/};
```
@z33ky This causes a segfault in the colour buffer code in main.cpp. It is seemingly cleared from the time it is created to the time it is needed in the game loop.. Could the compiler possibly be auto optimising this in???

Quaker762 commented:
@z33ky Uh-oh....

Aside from the conflict (which I thought would happen...), it's good to go! No more 2982839428374 number of elements to draw (and grinding the program to a halt haha).

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,120 @@
+        /**
+         *  Default constructor.
+         */
+        buffer_object(Target type = Target::ARRAY_BUFFER, const std::string& buffName = ""): buffType(type), name(buffName){Create();};
```
Slap a FIXME here, since we only set a default value for `type`, because we're not passing anything when creating the buffers for the VAO.
Look at my mail from the 17th for some suggestions. I like the last one the most, where we pair the SLOT-enum and the targets in a struct.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,120 @@
+         *  Give some data to our buffer
+         *
+         *  @param num Size (in bytes) of the data we are flushing.
+         *  @param data void pointer to our data (usually an std::array or std::vector).
```
`usage` docs are missing.
Reorder `data` before `num`.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,120 @@
+        void BufferData(void* data, std::size_t num, GLenum usage);
+
+        /**
+         *  Redefines/updates a subset of data stored in this GLBuffer.
```
Parameter docs are missing.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,120 @@
+        buffer_object(Target type = Target::ARRAY_BUFFER, const std::string& buffName = ""): buffType(type), name(buffName){Create();};
+
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted by the C++ Runtime.
```
See https://github.com/Palm-Studios/sh3redux/pull/94#discussion_r122571809.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+         */
+        void Bind();
+
+         /**
```
Incorrect indent.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+          */
+        void Unbind();
+
+          /**
```
Incorrect indent.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+         *
+         *  @param  offset - Offset into the buffer
+         *
+         *  @return void* - Pointer to value.
```
Type needs not be specified in doc.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+         *
+         *  @return void* - Pointer to value.
+         */
+        static constexpr void* buffer_offset(GLint offset){return static_cast<char *>(NULL) + offset;}
```
Use `nullptr` instead of `NULL`. The usage here is fine, but in other places `NULL` can select incorrect overloads since it's an integer in C++. So we should just generally avoid it.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+        Bind();
+        glEnableVertexAttribArray(attrib); // Set attrib index as currently being used
+        ASSERT_MSG(attrib < numbuffers, "VBO index (attribute) out of range");
+        buffers[attrib].Bind(); // Bind VBO we are going to send to
```
The code seems to clearly express what's being done without the comment.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+    void mutablevao<numbuffers>::SetDataLocation(GLuint attrib, DataType type, GLint size, GLsizei stride, GLint offset)
+    {
+        Bind();
+        glEnableVertexAttribArray(attrib); // Set attrib index as currently being used
```
The code seems to clearly express what's being done without the comment.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,311 @@
+    /**
+     *  A wrapper for an OpenGL 3.3 Vertex Array Object.
+     */
+    struct vertarray final
```
This can be removed now, right?

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,94 @@
+        glDrawArrays(GL_TRIANGLES, 0, vertices.GetNumberElements());
+}
+
+//void vertarray::Create()
```
Remove?

z33ky commented on source/SH3/system/glvertarray.cpp:
```diff
@@ -0,0 +1,94 @@
+#include "SH3/system/log.hpp"
+
+constexpr int   GL_VAO_UNBIND = 0;
+constexpr void* BUFFER_OFFSET(std::uint32_t i){return((char *)NULL + (i));}
```
remove

z33ky commented on source/main.cpp:
```diff
@@ -32,17 +54,46 @@ int main(int argc, char** argv)
+    sh3_gl::vao<MyAttributes> myVao;
+
+    myVao.Bind();
+    sh3_gl::buffer_object verts = myVao[MyAttributes::VERTEX_SLOT];
```
This can be a reference.
In fact I'm not sure if there's a good reason to have `buffer_object` and `vao` being copyable. We may just want to delete their copy ctor. If we find a use-case for it later, we can still change them to allow it.

z33ky reviewed (changes requested):
some `*vao` stuff is missing documentation.

The documentation for `vao::Attribute` should mention our assumptions, e.g. there's a `Attribute::MAX` and `Attribute::VERTEX_SLOT` and VAO using an IBO have an `Attribute::INDEX_SLOT`.
Note that Doxygen has a keyword to document template parameters: `tparam`.

For some parameters, you use `@param name - doc` - I'm not sure if Doxygen removes the dash.

It seems to me that the `glEnableVertexAttribArray` is missing a symmetric `glDisableVertexAttribArray` call, or is this not needed?

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,126 @@
+         *  Constructor.
+         *
+         *  FIXME: We're currently <i>always</i> passing @ref Target::ARRAY_BUFFER to the buffer object,
+         *  as we never actually pass anything to this constructor.
```
Note that we don't pass anything in `@ref mutablevao`.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,126 @@
+        buffer_object(Target type = Target::ARRAY_BUFFER, const std::string& buffName = ""): buffType(type), name(buffName){Create();};
+
+        /**
+         *  Destructor. Ensure we have no leaks when this buffer is to be deleted.
```
Destructor implies that memory gets freed. I think just "Destructor." suffices.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,126 @@
+         *
+         *  @param data void pointer to our data (usually an std::array or std::vector).
+         *  @param num Size (in bytes) of the data we are flushing.
+         *  @param usage - The way in which want this data stored (i.e, can we rewrite sections of it?).
```
Remove the dash.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,126 @@
+        /**
+         *  Give some data to our buffer
+         *
+         *  @param data void pointer to our data (usually an std::array or std::vector).
```
I'd just say "pointer to our data".

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,126 @@
+        /**
+         *  Redefines/updates a subset of data stored in this GLBuffer.
+         *
+         *  @param data void pointer to our data (usually an std::array or std::vector).
```
Same as for BufferData.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,249 @@
+    /**
+     *  Contains actual functionality of our VAO (@ref Bind, @ref Create etc).
+     *
+     *  @note Cannot be directly instantiated, use @ref vao instead
```
Can != should. This should not be directly instantiated.
Also, add a trailing dot.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,249 @@
+        static constexpr mutablevao<T::MAX> test(...);
+
+    public:
+        using type = decltype(test<Attribute>(0)); /**< Type of vao this is (either @ref mutablevao or @ref indexedmutablevao)
```
Missing `*/`.

z33ky commented on tests/vao.cpp:
```diff
@@ -0,0 +1,100 @@
+    MAX
+};
+
+/**
```
I don't think we need to Doxygen-comment our tests.
If we write our own testing framework, then that should be documented, but that's still in the future, if ever.

z33ky commented on tests/vao.cpp:
```diff
@@ -0,0 +1,100 @@
+    myVao.Bind(); // Bind our VAO
+	
+	// Create the vertex buffer object
+    sh3_gl::buffer_object verts = myVao[MyAttributes::VERTEX_SLOT];
```
Reference, as in main.cpp.

z33ky commented on tests/vao.cpp:
```diff
@@ -0,0 +1,100 @@
+    myVao.SetDataLocation(MyAttributes::VERTEX_SLOT, sh3_gl::vao<MyAttributes>::DataType::FLOAT, 3, 0, 0);
+
+	// Create the color buffer object
+    sh3_gl::buffer_object col = myVao[MyAttributes::COLOR_SLOT];
```
Reference, as in main.cpp.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,249 @@
+        //Prevent dynamic instantiation since we have no virtual dtor.
+        static void* operator new(std::size_t) = delete;
+        static void* operator new[](std::size_t) = delete;
+
```
Also delete the 'delete` operators.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,249 @@
+        /**
+         *  Indexed constructor.
+         */
+        finalvao(const buffer_object &ibo, GLsizei vertex_count, const buffer_object& verts):ibo(&ibo), vertices(verts)
```
Remove `vertex_count`.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,249 @@
+
+    protected:
+        GLuint  id;                         /**< Our VAO's identification number. **/
+        GLsizei vertex_count;               /**< The number of vertices to draw. **/
```
Remove.

z33ky commented on SILENT HILL 3 Redux.cbp:
```diff
@@ -134,10 +138,14 @@
+		<Unit filename="source/SH3/system/glbuffer.cpp" />
 		<Unit filename="source/SH3/system/glcontext.cpp" />
 		<Unit filename="source/SH3/system/glprogram.cpp" />
+		<Unit filename="source/SH3/system/glvertarray.cpp" />
```
Add those to `source/CMakeLists.txt` as well.

Quaker762 replied:
They're already in there, hence the merge conflict.

z33ky reviewed (changes requested):
Some *vao ctors are missing parameter or any documentation.
There's still some dashes following the parameter names in `@param` docs.

Sorry, this drags on so long, but after these superficial things are fixed I'll approve it. I promise ;)

Quaker762 commented:
> Sorry, this drags on so long, but after these superficial things are fixed I'll approve it. I promise ;)

Haha it's no issue, I'd rather have the code nice as well instead of being half arsed.

z33ky commented on include/SH3/system/glbuffer.hpp:
```diff
@@ -0,0 +1,127 @@
+         *
+         *  @param data void pointer to our data.
+         *  @param num Size (in bytes) of the data we are flushing.
+         *  @param usage - The way in which want this data stored (i.e, can we rewrite sections of it?).
```
Remove the dash here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Non indexed constructor.
+         */
+        finalvao(const buffer_object& verts):ibo(nullptr), vertices(verts)
```
@Param docs missing.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Indexed constructor.
+         */
+        finalvao(const buffer_object &ibo, const buffer_object& verts):ibo(&ibo), vertices(verts)
```
Param docs missing.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Set vertex shader location of data in VBO for specified attribute.
+         *
+         *  @param attrib - Attribute we want to map data input in our shader (using location = x).
```
Remove the dashes here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Overload the [] operator to get a buffer stored in @ref buffers more easily.
+         *
+         *  @param slot - Index slot where the buffer is stored. Usually in a @ref Attribute.
```
Remove the dash here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Get offset (stride) into a buffer stored in @ref buffers
+         *
+         *  @param offset - Offset to the first element.
```
Remove the dash here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+         *
+         *  @param offset - Offset to the first element.
+         *
+         *  @return void* - Pointer to the first element of stride 'offset'.
```
And here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+    /**
+     *  Set vertex shader location of data in VBO for specified attribute.
+     *
+     *  @tparam numbuffers - Number of buffers
```
Remove dashes here.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+    struct indexedmutablevao : public mutablevao<numbuffers>
+    {
+    protected:
+        indexedmutablevao(const buffer_object& vertices): mutablevao<numbuffers>(mutablevao<numbuffers>::buffers[index_slot], numbuffers, vertices) {}
```
Documentation missing here. Don't forget the parameter ;)

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+     *
+     *  Either @ref indexedmutablevao or @ref mutablevao (unindexed).
+     *
+     *  @tparam Attribute - Attribute enum (specifies Attribute::MAX to send to @ref mutablevao).
```
Remove the dash here.

z33ky replied:
Also document the usage of `Attribute::INDEX_SLOT`.

I'd just leave a short parameter description for `@tparam` and put the description of it below:
```
@tparam Attribute An attribute enum.

@p Attribute should have a value @c MAX to indicate the number of attributes (slots).
If @p Attribute has a value @c INDEX_SLOT, then the @ref type is an @ref indexedmutablevao.
```

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+    };
+
+    template<typename Attribute>
+    struct vao final : public vaoparent<Attribute>::type
```
Missing docs.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,250 @@
+        /**
+         *  Set vertex shader location of data in VBO for specified attribute.
+         *
+         *  @param attrib - Attribute we want to map data input in our shader (using location = x).
```
Remove the dashes.

z33ky reviewed (changes requested):
Did you test whether Doxygen ignores the dashes after parameter names?
I now marked all places where this occurs.

In my code I aligned the parameter documentation, which also visually stands out, which I assume is why you used the dashes.
```
@param data  Void pointer to our data.
@param num   Size (in bytes) of the data we are flushing.
@param usage The way in which want this data stored (i.e, can we rewrite sections of it?).
```

I also pointed out where exactly documentation is missing.

z33ky commented:
Also, to resolve the merge conflicts you should rebase your code on master.

Quaker762 commented:
> Did you test whether Doxygen ignores the dashes after parameter names?

I just checked, it keeps them in there. I'm just used to using dashes in a list of things, but aligning is a lot cleaner. I'll remove all of them.

> I also pointed out where exactly documentation is missing.

Whoops, I must have missed some of that sorry.

Quaker762 committed:
Abstraction of OpenGL Data Types

All important OpenGL structures and data types have been abstracted to
make them more memory safe and remove the need for a whole lot of
repetitive 'glXYZ()' calls.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+     */
+    template<std::size_t numbuffers, std::size_t index_slot>
+
+    /**
```
This should be moved to the ctor.
Also, VAo -> VAO.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+    template<std::size_t numbuffers, std::size_t index_slot>
+
+    /**
+     *  VAo that inherits from @ref mutablevao. This passes our index @ref buffer_object to the @ref finalvao cosntructor.
```
Too implementation specific. Just "Constructor" is fine.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+    /**
+     *  VAo that inherits from @ref mutablevao. This passes our index @ref buffer_object to the @ref finalvao cosntructor.
+     *
+     *  @param vertices Reference to our vertices @ref buffer_object. This gets passed to @ref finalvao.
```
" This gets passed to @ref finalvao.": unnecessary implementation detail.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+     *
+     *  Either @ref indexedmutablevao or @ref mutablevao (unindexed).
+     *
+     *  @tparam Attribute Attribute enum (specifies Attribute::MAX to send to @ref mutablevao).
```
See https://github.com/Palm-Studios/sh3redux/pull/94#discussion_r128913919.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+     *  Our VAO object. This handles all @ref buffer_objects needed to draw data that has been
+     *  sent to the graphics card.
+     *
+     *  @tparam Attribute The attribute enum that contains all slots required by this VAO. This
```
This should also mention `VERTEX_SLOT`, `MAX` and optional `INDEX_SLOT`.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+     *  sent to the graphics card.
+     *
+     *  @tparam Attribute The attribute enum that contains all slots required by this VAO. This
+     *  makes it more generic, i.e, some VAOs might need a color buffer instead of a UV buffer etc.
```
I'd indent it so it looks like
```
     *  @tparam Attribute The attribute enum that ...
     *                    makes it more generic, ...
```

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,287 @@
+        /**
+         *  Our VAO that inherits from @ref finalvao. Contains a list of all our @ref buffer_object structures.
+         *
+         *  @param vertices Reference to our vertices @ref buffer_object. This gets passed
```
" This gets passed to @ref": this sentence is missing an end. Also, unnecessary implementation detail. Just omit it.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,291 @@
+     *  @tparam index_slot  The slot number where our index @ref buffer_object resides.
+     */
+    template<std::size_t numbuffers, std::size_t index_slot>
+
```
Remove this empty line.

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,291 @@
+     *  Our VAO object. This handles all @ref buffer_objects needed to draw data that has been
+     *  sent to the graphics card.
+     *
+     *  @tparam Attribute   The attribute enum that contains all slots required by this VAO. This
```
This is a very long parameter description. Do something like the `Attribute` doc for `vaoparent`.

z33ky commented:
Ah, and don't forget to use `@c` and `@p` like for the `Attribute` of `vaoparent`. It instructs Doxygen to use typewriter font.

Quaker762 commented:
> Ah, and don't forget to use @c and @p like for the Attribute of vaoparent.

Where did you want me to put that?? `@tparam @p Attribute`???

z33ky commented:
No. See https://github.com/Quaker762/sh3redux/blob/8071b83685bfcaaa0686e5deb076b249edbf8c1f/include/SH3/system/glvertarray.hpp#L232. Use `@c` to refer to values (including symbolic ones like enum members) and `@p` to refer to the parameter (outside of `@param Attribute`).
For forget what the difference between the two commandos is, but it made sense this way when I looked them up.

Quaker762 commented:
@z33ky This should be all good to go unless I've missed anything else (which I hope I haven't haha)

z33ky commented on include/SH3/system/glvertarray.hpp:
```diff
@@ -0,0 +1,288 @@
+     *  Our VAO object. This handles all @ref buffer_objects needed to draw data that has been.
+     *  sent to the graphics card.
+     *
+     *  @tparam Attribute   @p Attribute enum that contains all slots required by this VAO, making it
```
This is still a very long parameter description.
The idea is to be brief here and go into detail below.
Look how `Attribute` is documented for `vaoparent`.

Quaker762 commented:
Now we're rolling ;)
